src/renderer/d3d.rs

```diff
@@ -10,186 +7,122 @@ use skia_safe::{
+use windows::Win32::Graphics::Direct3D::D3D_FEATURE_LEVEL_11_0;
+use windows::Win32::Graphics::Direct3D12::{
+    D3D12CreateDevice, D3D12GetDebugInterface, ID3D12CommandQueue, ID3D12Debug, ID3D12Device,
+    ID3D12Fence, ID3D12Resource, D3D12_COMMAND_LIST_TYPE_DIRECT, D3D12_COMMAND_QUEUE_DESC,
+    D3D12_COMMAND_QUEUE_FLAG_NONE, D3D12_FENCE_FLAG_NONE, D3D12_RESOURCE_STATE_PRESENT,
+};
+use windows::Win32::Graphics::Dxgi::Common::{
+    DXGI_ALPHA_MODE_UNSPECIFIED, DXGI_FORMAT_R8G8B8A8_UNORM, DXGI_SAMPLE_DESC,
+};
+use windows::Win32::Graphics::Dxgi::{
+    CreateDXGIFactory2, IDXGIAdapter1, IDXGIFactory4, IDXGISwapChain1, IDXGISwapChain3,
+    DXGI_ADAPTER_FLAG, DXGI_ADAPTER_FLAG_NONE, DXGI_ADAPTER_FLAG_SOFTWARE,
+    DXGI_CREATE_FACTORY_DEBUG, DXGI_SCALING_NONE, DXGI_SWAP_CHAIN_DESC1,
+    DXGI_SWAP_CHAIN_FLAG_FRAME_LATENCY_WAITABLE_OBJECT, DXGI_SWAP_EFFECT_FLIP_DISCARD,
+    DXGI_USAGE_RENDER_TARGET_OUTPUT,
+};
+use windows::Win32::System::Threading::{CreateEventW, WaitForSingleObjectEx, INFINITE};
+use winit::{
+    event_loop::EventLoopProxy,
+    raw_window_handle::{HasWindowHandle, RawWindowHandle},
+    window::Window,
 };
-use winit::{event_loop::EventLoopProxy, window::Window};
-use wio::com::ComPtr;
 
 use super::{vsync::VSyncWinSwapChain, SkiaRenderer, VSync};
 #[cfg(feature = "gpu_profiling")]
 use crate::profiling::{d3d::create_d3d_gpu_context, GpuCtx};
 use crate::{profiling::tracy_gpu_zone, window::UserEvent};
 
-const D3D_FEATUREL_LEVEL: D3D_FEATURE_LEVEL = D3D_FEATURE_LEVEL_11_0;
-
-pub fn call_com_fn<T0, T1, F>(fun: F) -> Result<ComPtr<T1>, ()>
-where
-    T1: Interface,
-    F: FnOnce(&mut *mut T0, REFIID) -> HRESULT,
-{
-    let mut ptr = null_mut();
-    let res = fun(&mut ptr, &T1::uuidof());
-    if SUCCEEDED(res) {
-        Ok(unsafe { ComPtr::from_raw(ptr as *mut T1) })
-    } else {
-        Err(())
-    }
-}
+fn get_hardware_adapter(factory: &IDXGIFactory4) -> Result<IDXGIAdapter1> {
+    for i in 0.. {
+        let adapter = unsafe { factory.EnumAdapters1(i)? };
+        let mut desc = Default::default();
+        unsafe {
+            adapter.GetDesc1(&mut desc)?;
+        }
 
-fn is_adapter_suitable(adapter: *mut IDXGIAdapter1) -> bool {
-    let mut desc = DXGI_ADAPTER_DESC1::default();
-    if SUCCEEDED(unsafe { (*adapter).GetDesc1(&mut desc) }) {
-        if (desc.Flags & DXGI_ADAPTER_FLAG_SOFTWARE) == DXGI_ADAPTER_FLAG_SOFTWARE {
-            // Don't select the Basic Render Driver adapter.
-            false
-        } else {
-            // Check to see whether the adapter supports Direct3D 12, but don't create the
-            // actual device yet.
-            unsafe {
-                SUCCEEDED(D3D12CreateDevice(
-                    adapter as *mut IUnknown,
-                    D3D_FEATUREL_LEVEL,
-                    &ID3D12Device::uuidof(),
-                    null_mut(),
-                ))
-            }
+        if (DXGI_ADAPTER_FLAG(desc.Flags as i32) & DXGI_ADAPTER_FLAG_SOFTWARE)
```

Comment:
I think this can be written as `DXGI_ADAPTER_FLAG(desc.Flags as i32).contains(DXGI_ADAPTER_FLAG_SOFTWARE)`. And there are probably other similar usages that can be changed.